hw/mcu/nxp/kinetis/src/hal_spi.c

```diff
@@ -0,0 +1,623 @@
+    }
+
+    if (spi_type == HAL_SPI_TYPE_MASTER) {
+        return hal_spi_init_master(spi, cfg);
+    } else {
+        return hal_spi_init_slave(spi, cfg);
+    }
+    return EINVAL;
+}
+
+int
+hal_spi_init_hw(uint8_t spi_num,
+                uint8_t spi_type,
+                const struct hal_spi_hw_settings *cfg)
+{
+    struct nxp_hal_spi_cfg hal_cfg;
+    hal_cfg.clk_pin = (uint8_t)cfg->pin_sck;
+    if (spi_type == HAL_SPI_TYPE_MASTER) {
+        hal_cfg.sout_pin = (uint8_t)cfg->pin_mosi;
+        hal_cfg.sin_pin = (uint8_t)cfg->pin_miso;
+    } else {
+        hal_cfg.sin_pin = (uint8_t)cfg->pin_mosi;
+        hal_cfg.sout_pin = (uint8_t)cfg->pin_miso;
+    }
+    hal_cfg.pcs_pin = (uint8_t)cfg->pin_ss;
+    return hal_spi_init(spi_num, &hal_cfg, spi_type);
+}
+
+int
+hal_spi_config(int spi_num, struct hal_spi_settings *settings)
+{
+    struct nxp_hal_spi *spi;
+    struct nxp_spi_master *master;
+    struct nxp_spi_slave *slave;
+    dspi_clock_polarity_t cpol;
+    dspi_clock_phase_t cpha;
+
+    spi = hal_spi_resolve(spi_num);
+    if (!spi) {
+        return EINVAL;
+    }
+
+    if (!settings) {
+        return EINVAL;
+    }
+
+    switch (settings->data_mode) {
+    case HAL_SPI_MODE0:
+        cpol = kDSPI_ClockPolarityActiveHigh;
+        cpha = kDSPI_ClockPhaseFirstEdge;
+        break;
+    case HAL_SPI_MODE1:
+        cpol = kDSPI_ClockPolarityActiveHigh;
+        cpha = kDSPI_ClockPhaseSecondEdge;
+        break;
+    case HAL_SPI_MODE2:
+        cpol = kDSPI_ClockPolarityActiveLow;
+        cpha = kDSPI_ClockPhaseFirstEdge;
+        break;
+    case HAL_SPI_MODE3:
+        cpol = kDSPI_ClockPolarityActiveLow;
+        cpha = kDSPI_ClockPhaseSecondEdge;
+        break;
+    default:
+        assert(0);
+        break;
+    }
+
+    if (spi->type == HAL_SPI_TYPE_MASTER) {
+        master = (struct nxp_spi_master *) spi;
+
+        master->config.ctarConfig.baudRate = settings->baudrate;
+        master->config.ctarConfig.pcsToSckDelayInNanoSec =
+            1000000000U / master->config.ctarConfig.baudRate;
+        master->config.ctarConfig.lastSckToPcsDelayInNanoSec =
+            1000000000U / master->config.ctarConfig.baudRate;
+        master->config.ctarConfig.betweenTransferDelayInNanoSec =
+            1000000000U / master->config.ctarConfig.baudRate;
+        master->config.ctarConfig.direction =
+            (settings->data_order == HAL_SPI_MSB_FIRST) ?
+            kDSPI_MsbFirst :
+            kDSPI_LsbFirst;
+        master->config.ctarConfig.bitsPerFrame =
+            (settings->word_size == HAL_SPI_WORD_SIZE_8BIT) ? 8 : 9;
+        master->config.ctarConfig.cpol = cpol;
+        master->config.ctarConfig.cpha = cpha;
+    } else {
+        slave = (struct nxp_spi_slave *) spi;
+        slave->config.ctarConfig.bitsPerFrame =
+            (settings->word_size == HAL_SPI_WORD_SIZE_8BIT) ? 8 : 9;
+        slave->config.ctarConfig.cpol = cpol;
+        slave->config.ctarConfig.cpha = cpha;
+    }
```

Comment:
Since `nxp_spi_master` and `nxp_spi_slave` contain the `nxp_hal_spi` as the first element you could probably just do `master` specific configuration inside the `if` and move the generic after the `if` block:

```
spi->config.ctarConfig.bitsPerFrame =
    (settings->word_size == HAL_SPI_WORD_SIZE_8BIT) ? 8 : 9;
spi->config.ctarConfig.cpol = cpol;
spi->config.ctarConfig.cpha = cpha;
```

And get rid of the `slave` variable. Not entirely sure, disregard if you think it won't work.

Reply:
Unfortunately `ctarConfig` is `dspi_master_ctar_config_t` on `dspi_master_config_t` and `dspi_slave_ctar_config_t` on `dspi_slave_config_t`